### ISSUE TITLE
reinterpret_cast-ing nonOwnedArrayDeleter causes failures in Clang's Undefined Behavior Sanitizer

### ISSUE THREAD

**Issue by CaseyCarlin** [mosra/magnum#531]
There are a few places in `Magnum::Trade` where nonOwnedArrayDeleter is reinterpret_cast-ed to a different function signature to meet the requirements of `Array`'s deleter:
```
Containers::Array<MeshAttributeData> meshAttributeDataNonOwningArray(const Containers::ArrayView<const MeshAttributeData> view) {
    /* Ugly, eh? */
    return Containers::Array<MeshAttributeData>{const_cast<MeshAttributeData*>(view.data()), view.size(), reinterpret_cast<void(*)(MeshAttributeData*, std::size_t)>(Implementation::nonOwnedArrayDeleter)};
}
```
This sort of shenanigans triggers the `function-type-mismatch` check in ubsan once the deleter is called in Array::~Array. The function signature is different at the call site: `void(char*, unsigned long)` called as `void (Magnum::Trade::MeshAttributeData*, unsigned long)` 

I can put together a repro if needed, but I found this trivially with code that populates a MeshData instance and runs it through `MeshTools::owned`. I am building with clang 10 and `-fsanitize=undefined`

`nonOwnedArrayDeleter` is a no-op so this does not _really_ result in undefined behavior, but I am building in a fairly zero-tolerance environment. I will disable this check as narrowly as possible, but being able to use Magnum/Corrade unmodified with a happy sanitizer is my preferred outcome here.

**Comment by mosra** [comment]
Hi,

first of all -- sorry for a very late reply. I'm only slowly going though the backlog accumulated during the summer.

I'm aware of this being a "mildly out-of-ordinary" code (that's why the comment there), but since the function is a no-op I don't expect any real-world problems caused by it. I chose to do it this way instead of having to export a handful of no-op deleters with slightly different signatures to satisfy the type checker, saving *slightly* on the count of exported symbols, but that was maybe not the best idea. 

What do you suggest as a solution?

- Is it possible to suppress the UBSan warning for this particular function using some `#pragma`?
- Would some C-style variadic (`...`) function work in this case? I'd still need to cast it to the deleter type, so that's probably solving nothing.
- Or just give up and stamp out a different deleter for each type I use it for?

**Comment by mosra** [comment]
Even after such a long time (sorry again) I didn't manage to find anything better/cleaner than the third suggestion, so I implemented that. This should be fixed as of 001be98a88e5e32adc1429cea6f6a44df17200e6 and I'll try to not make similar UB-triggering hacks in the future :)
